src/lib_json/json_reader.cpp

```diff
@@ -1991,6 +1978,32 @@ bool parseFromStream(CharReader::Factory const& fact, IStream& sin, Value* root,
   return reader->parse(begin, end, root, errs);
 }
 
+Json::LineAndColumn getLocationLineAndColumn(char const* beginDoc,
+                                             char const* endDoc,
+                                             size_t location) {
+  Json::LineAndColumn lineAndColumn;
+  char const* current = beginDoc;
+  char const* lastLineStart = current;
+  int line = 0;
+  while (current < beginDoc + location && current < endDoc) {
+    char c = *current++;
+    if (c == '\r') {
+      if (*current == '\n')
+        ++current;
+      lastLineStart = current;
+      ++line;
+    } else if (c == '\n') {
+      lastLineStart = current;
+      ++line;
+    }
+  }
+  // column & line start at 1
+  lineAndColumn.column = int(beginDoc + location - lastLineStart) + 1;
+  lineAndColumn.line = ++line;
```

Comment:
I think this function can be written more simply than this.
You don't need to introduce variables for `lineAndColumn`, `current`, or `lastLineStart`.

You don't need `endDoc`. It's not handled correctly anyway.
If the `location` indicates a position beyond `endDoc`, the result is silently considered to be a position in an imaginary last line of infinite length. This is not a good failure mode and so don't even try, I think.

```c++
    int line = 1;
    int col = 1;
    const char* last = beginDoc + location;
    for (; beginDoc != last; ++beginDoc) {
      switch (*beginDoc) {
        case '\r':
          if (beginDoc + 1 != last && beginDoc[1] == '\n')
            continue;  // consume CRLF as a single token.
          [[fallthrough]];  // CR without a following LF is same as LF
        case '\n':
          col = 1;
          ++line;
          break;
        default:
          ++col;
          break;
      }
    }
    return {line, col};
```